cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll

```diff
@@ -19,9 +19,21 @@ class ValueNumberPropertyProvider extends IRPropertyProvider {
  * The value number assigned to a particular set of instructions that produce equivalent results.
  */
 class ValueNumber extends TValueNumber {
-  final string toString() { result = getExampleInstruction().getResultId() }
+  final string toString() { result = "GVN" }
 
-  final Language::Location getLocation() { result = getExampleInstruction().getLocation() }
```

Comment:
Can't we just implement `hasLocationInfo` for `ValueNumberBound` by forwarding the call to `ValueNumber::hasLocationInfo` (and forget about `getLocation`)?

I'm not exactly sure about the implications of this, but looking at the documentation for `hasLocationInfo` it seems to be the more efficient way to get a access to a location

Reply:
I understand what you mean on the "earliest non-unknown location" now. I haven't solved this yet (still trying to figure out how to specify the earliest location in QL), but I pushed the changes in .expected anyway even though the test reports too many locations for each Bound. Hopefully this will make the PR green!

Reply:
What motivates the switch from `getLocation` to `hasLocationInfo` ? Is it https://github.com/Semmle/ql/pull/1542#issuecomment-509545966? I must admit I've forgotten why I wrote that comment.

In any case, what need is there to change the locations/strings of the `Instruction`-based API in a PR that's about adding an `Expr`-based API? Does the latter depend on the former?

Reply:
> What motivates the switch from `getLocation` to `hasLocationInfo` ? Is it [#1542 (comment)](https://github.com/Semmle/ql/pull/1542#issuecomment-509545966)? I must admit I've forgotten why I wrote that comment.

The main motivation for switching away from the `getLocation` (I think) was the following:

> These are currently implemented in terms of getExampleInstruction, but getExampleInstruction is an expensive operation that uses Instruction.getUniqueId, whose QLDoc says "This is used for sorting IR output for tests, and is likely to be inefficient for any other use." We don't want to invoke such expensive operations implicitly.

But I guess we can keep using `getLocation`, but implement it without a call to `getExampleInstruction`

Reply:
Yes, that will do.